### PR TITLE
Convert to python package

### PR DESCRIPTION
**Pull request by camfairchild**
@eduardogr

### CODE REVIEW

File: python/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+[project]
+name = "subtensornodeapi"
+version = "0.0.1"
```
Comment (eduardogr) [inline python/pyproject.toml]: can you get this from VERSION, or at least align it

File: python/setup.py [hunk 2]
```diff
@@ -39,24 +39,4 @@
     # include neither binaries
     pass
 
-setup(
```
Comment (eduardogr) [inline python/setup.py]: note that you have more things to setup in the current version of the pythonize branch

File: python/src/subtensornodeapi/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+__version__ = '0.0.1'
```
Comment (eduardogr) [inline python/src/subtensornodeapi/__init__.py]: samething, let's centralized where we have the VERSION. And change the update_version script aligned with that

File: subtensor-api/pyproject.toml [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
-version = "0.0.1"
 authors = [
   { name="Opentensor Foundation", email="cameron@opentensor.ai" },
   { name="Opentensor Foundation", email="eduardo@opentensor.ai" },
 ]
-license = "MIT"
+license = { file = "LICENSE" }
 description = "Subtensor API package for bittensor"
 readme = "README.md"
+dynamic = ["version"]
```
Comment (eduardogr) [inline subtensor-api/pyproject.toml:14]: nice!

File: subtensor-api/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+setuptools
+tqdm
```
Comment (eduardogr) [inline subtensor-api/requirements.txt]: if you can pin some of their versions would be nice

File: subtensor-api/setup.py [hunk 6]
```diff
@@ -38,4 +41,10 @@
     # neither linux or macos
     raise Exception("Unsupported platform: {}".format(platform))
 
-setup()
+setup(
+    extras_require={
+        'test': ['unittest'],
+        'dev': ['build'],
```
Comment (eduardogr) [inline subtensor-api/setup.py]: what is this?

File: subtensor-api/src/subtensorapi/cli.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
+# The MIT License (MIT)
```
Comment (eduardogr) [inline subtensor-api/src/subtensorapi/cli.py:1]: Check the path of these files.

I think it should be `subtensor-api/subtensorapi/cli-py`, Is that right?

File: subtensor-api/src/subtensorapi/cli.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+# Copyright © 2022 Opentensor Foundation
```
Comment (eduardogr) [inline subtensor-api/src/subtensorapi/cli.py:2]: also, what about renaming this file to something different, this is not going to be a CLI, right?

File: subtensor-api/tests/test_fast_sync.py [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+                                       FastSyncFormatException,
+                                       FastSyncNotFoundException,
+                                       FastSyncOSNotSupportedException,
+                                       FastSyncRuntimeException)
+
+U64MAX = 18446744073709551615
+U32MAX = 4294967295
+RAOPERTAO = 10e9
+
+class TestLoadNeurons(unittest.TestCase):
```
Comment (eduardogr) [inline subtensor-api/tests/test_fast_sync.py:19]: nice!

**Comment by camfairchild** [comment]
@eduardogr I think it should be good to go now
I was able to get it to work on Ubuntu 18.04 with Python 3.7 (in Docker)

**Comment by camfairchild** [comment]
It also already works on Morpheus (Ubuntu 20.04, Python 3.8,3.9, 3.10)

File: run_test.sh [hunk 10]
```diff
@@ -0,0 +1 @@
+python3 -m subtensornodeapi sync_and_save -f "./test.json"
```
Comment (eduardogr) [inline run_test.sh]: do we still need this?

File: js/test/test.js [hunk 11]
```diff
@@ -0,0 +1,146 @@
+const {sync, get_block_at_registration_for_all} = require("../main");
+const {expect} = require("chai");
+
+describe("Test Neurons are pulled", function() {
+        const mock_n = 1024;
+
+        const get_mock_api = (n_neurons) => {
+            return {
```
Comment (eduardogr) [inline js/test/test.js:8]: Since this object is reused, what do you think about extracting it to use a var or const intead of copying it?

Comment (eduardogr) [reply js/test/test.js:8]: Or maybe better extracting `get_mock_api` to a common place, right?

File: subtensor-api/VERSION [hunk 12]
```diff
@@ -0,0 +1 @@
+0.0.13
```
Comment (eduardogr) [inline subtensor-api/VERSION]: going to 0.1.0 to release?

File: subtensor-api/pyproject.toml [hunk 13]
```diff
@@ -0,0 +1,41 @@
+[build-system]
+requires = ["setuptools>=61.0,<62.0", "build>=0.8.0,<1"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "subtensorapi"
+authors = [
+  { name="Opentensor Foundation", email="cameron@opentensor.ai" },
+  { name="Opentensor Foundation", email="eduardo@opentensor.ai" },
```
Comment (eduardogr) [inline subtensor-api/pyproject.toml:9]: thanks man! but it's basically your job! I appreciate the credits though

File: subtensor-api/pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+requires-python = ">=3.7"
+
+[tool.setuptools.dynamic]
+version = {file = "VERSION"}
+
+[project.urls]
+Documentation = "https://github.com/opentensor/subtensor-node-api#subtensor-node-api"
+Source = "https://github.com/opentensor/subtensor-node-api"
+
```
Comment (eduardogr) [inline subtensor-api/pyproject.toml]: extra line?

File: subtensor-api/tests/test_fast_sync.py [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+                    FastSync.verify_binary_exists()
+        
+    def test_binary_found(self):
+        with patch("os.path.exists", return_value=True):
+            with patch("os.path.isfile", return_value=True):
+                FastSync.verify_binary_exists() # no exception should be raised
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (eduardogr) [inline subtensor-api/tests/test_fast_sync.py:172]: Why do we need this?
Is this needed when using unittest `pytest? I woul say we can avoid this?

File: subtensor-api/src/subtensorapi/__main__.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    )
+    cli.add_args_blockAtRegistration_for_all_and_save(
+        blockAtRegistration_for_all_and_save
+    )
+
+    parsed_args = parser.parse_args(args)
+
+    if parsed_args.command == "sync_and_save":
+        cli.sync_and_save(parsed_args)
+    elif parsed_args.command == "blockAtRegistration_for_all_and_save":
```
Comment (eduardogr) [inline subtensor-api/src/subtensorapi/__main__.py]: is this not a very long command for a cli?
Could we have something similar with a shorter command?

**Review by eduardogr — CHANGES_REQUESTED** [review]
What about moving the binaries to a `bin/` directory?

File: subtensor-api/src/subtensorapi/__init__.py [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+            
+            for neuron_data in tqdm(data, "Parsing Neuron Data"):
+                # validate the neuron data
+                neuron = SimpleNamespace( **neuron_data )
+                # hotkey and coldkey are strings
+                # uid is an int
+                # active is an int
+                # ip is a string
+                # ip_type is an int
+                # port is an int
```
Comment (eduardogr) [inline subtensor-api/src/subtensorapi/__init__.py:282]: Why do we have these comments? where are those fields?

Comment (camfairchild) [reply subtensor-api/src/subtensorapi/__init__.py:282]: Those fields are added in the keyword deconstruction `neuron_data**`

I added the comments to make it clear that they're already parsed properly from the JSON

File: subtensor-api/tests/test_fast_sync.py (hunk 15 above)
Comment (camfairchild) [reply subtensor-api/tests/test_fast_sync.py:172]: this lets you run separate unittest testcases using the class name and the function, etc.

File: js/test/test.js (hunk 11 above)
Comment (camfairchild) [reply js/test/test.js:8]: they call different storage (last method in chain), but I could extract to the same API.

I did it separate as to only mock what SHOULD be used

File: subtensor-api/src/subtensorapi/__init__.py (hunk 17 above)
Comment (eduardogr) [reply subtensor-api/src/subtensorapi/__init__.py:282]: mmm it wasn't that clear to me. Do you think we can improve that commenting that the line before the deconstruaction? We could mention that and adding all the files enumerating them, what do you think about that?

**Review by eduardogr — APPROVED** [review]
LGTM

Nice job @camfairchild !

**Pull request merged**